File: test-scripts/send-test-messages/src/main/java/com/example/AsyncFlow/test/SendTestMessages.java
package com.example.AsyncFlow.test;

import com.example.AsyncFlow.test.model.*;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.sqs.SqsClient;
import software.amazon.awssdk.services.sqs.model.*;

import java.math.BigDecimal;
import java.net.URI;
import java.time.LocalDateTime;
import java.util.Base64;
import java.util.List;

/**
 * ============================================================
 *  AsyncFlow - Script de Teste (Java)
 * ============================================================
 *  Envia mensagens de teste para as filas SQS do LocalStack.
 *
 *  Pré-requisitos:
 *    - Docker rodando: docker-compose up -d
 *    - Maven instalado
 *
 *  Como executar:
 *    cd test-scripts/send-test-messages
 *    mvn package -q
 *    java -jar target/send-test-messages-1.0.0.jar
 * ============================================================
 */
public class SendTestMessages {

    // ── Configuração LocalStack ──────────────────────────────
    private static final String LOCALSTACK_ENDPOINT = "http://localhost:4566";
    private static final String ACCOUNT_ID          = "000000000000";
    private static final Region REGION              = Region.US_EAST_1;

    // ── Nomes das filas ──────────────────────────────────────
    private static final String QUEUE_VALIDATION  = "validacao-vendas";
    private static final String QUEUE_PERSISTENCE = "notas-processadas";
    private static final String QUEUE_MONITOR     = "fila-monitor";

    // ── Dados de teste ───────────────────────────────────────
    private static final ClientDto CLIENTE_1 = new ClientDto("12345678901", "Joao da Silva");
    private static final ClientDto CLIENTE_2 = new ClientDto("98765432100", "Maria Oliveira");

    private static final List<ItemDto> ITENS_1 = List.of(
            new ItemDto("SN-001", "Notebook Dell",    new BigDecimal("3500.00")),
            new ItemDto("SN-002", "Mouse Logitech",   new BigDecimal("150.00"))
    );
    private static final List<ItemDto> ITENS_2 = List.of(
            new ItemDto("SN-003", "Monitor LG 27",    new BigDecimal("1200.00")),
            new ItemDto("SN-004", "Teclado Mecanico", new BigDecimal("450.00")),
            new ItemDto("SN-005", "Headset Sony",     new BigDecimal("320.00"))
    );

    // XML simulado de NF
    private static final String XML_NF = """
            <?xml version="1.0" encoding="UTF-8"?>
            <NotaFiscal>
              <numero>NF-2026-002</numero>
              <emitente>
                <cnpj>11222333000181</cnpj>
                <nome>Empresa Teste LTDA</nome>
              </emitente>
              <destinatario>
                <cpf>98765432100</cpf>
                <nome>Maria Oliveira</nome>
              </destinatario>
              <itens>
                <item><serie>SN-003</serie><descricao>Monitor LG 27</descricao><valor>1200.00</valor></item>
                <item><serie>SN-004</serie><descricao>Teclado Mecanico</descricao><valor>450.00</valor></item>
                <item><serie>SN-005</serie><descricao>Headset Sony</descricao><valor>320.00</valor></item>
              </itens>
              <total>1970.00</total>
            </NotaFiscal>""";

    // ────────────────────────────────────────────────────────
    public static void main(String[] args) throws Exception {
        System.out.println("=".repeat(60));
        System.out.println("  AsyncFlow - Script de Teste de Mensagens SQS (Java)");
        System.out.println("=".repeat(60));

        // Configura Jackson com suporte a LocalDateTime
        ObjectMapper mapper = new ObjectMapper();
        mapper.registerModule(new JavaTimeModule());
        mapper.disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);

        // Cria cliente SQS apontando para LocalStack
        SqsClient sqs = SqsClient.builder()
                .endpointOverride(URI.create(LOCALSTACK_ENDPOINT))
                .region(REGION)
                .credentialsProvider(StaticCredentialsProvider.create(
                        AwsBasicCredentials.create("test", "test")))
                .build();

        // 1. Garante que as filas existem
        System.out.println("\n[1] Verificando/criando filas no LocalStack...");
        String urlValidation  = getOrCreateQueue(sqs, QUEUE_VALIDATION);
        String urlPersistence = getOrCreateQueue(sqs, QUEUE_PERSISTENCE);
        String urlMonitor     = getOrCreateQueue(sqs, QUEUE_MONITOR);

        // 2. Mensagens para o VALIDATION-WORKER
        System.out.println("\n[2] Enviando mensagens para o VALIDATION-WORKER...");

        NotaFiscalDto nf1Validation = new NotaFiscalDto(
                "NF-2026-001", CLIENTE_1, new BigDecimal("3650.00"),
                ITENS_1, LocalDateTime.now(), "11222333000181",
                "PROCESSANDO", null, null
        );
        sendMessage(sqs, urlValidation, new MessageRequest(nf1Validation), mapper,
                "NF-2026-001 (sem arquivo) -> validation-worker");

        String xmlBase64 = Base64.getEncoder().encodeToString(XML_NF.getBytes());
        NotaFiscalDto nf2Validation = new NotaFiscalDto(
                "NF-2026-002", CLIENTE_2, new BigDecimal("1970.00"),
                ITENS_2, LocalDateTime.now(), "11222333000181",
                "PROCESSANDO", xmlBase64, "nf-2026-002.xml"
        );
        sendMessage(sqs, urlValidation, new MessageRequest(nf2Validation), mapper,
                "NF-2026-002 (com XML)     -> validation-worker");

        // 3. Mensagens para o PERSISTENCE-WORKER
        System.out.println("\n[3] Enviando mensagens para o PERSISTENCE-WORKER...");

        NotaFiscalDto nf1Persistence = new NotaFiscalDto(
                "NF-2026-001", CLIENTE_1, new BigDecimal("3650.00"),
                ITENS_1, LocalDateTime.now(), "11222333000181",
                "VALIDADA", null, null
        );
        sendMessage(sqs, urlPersistence, new MessageRequest(nf1Persistence), mapper,
                "NF-2026-001 (sem arquivo) -> persistence-worker");

        NotaFiscalDto nf2Persistence = new NotaFiscalDto(
                "NF-2026-002", CLIENTE_2, new BigDecimal("1970.00"),
                ITENS_2, LocalDateTime.now(), "11222333000181",
                "VALIDADA", xmlBase64, "nf-2026-002.xml"
        );
        sendMessage(sqs, urlPersistence, new MessageRequest(nf2Persistence), mapper,
                "NF-2026-002 (com XML)     -> persistence-worker");

        System.out.println("\n" + "=".repeat(60));
        System.out.println("  Mensagens enviadas com sucesso!");
        System.out.println("  Acesse o dashboard: http://localhost:8083");
        System.out.println("  Swagger monitor:    http://localhost:8083/swagger-ui.html");
        System.out.println("=".repeat(60));

        sqs.close();
    }

    // ── Helpers ──────────────────────────────────────────────

    private static String getOrCreateQueue(SqsClient sqs, String queueName) {
        String queueUrl = buildQueueUrl(queueName);
        try {
            sqs.getQueueAttributes(GetQueueAttributesRequest.builder()
                    .queueUrl(queueUrl)
                    .attributeNames(QueueAttributeName.APPROXIMATE_NUMBER_OF_MESSAGES)
                    .build());
            System.out.println("  [OK]     Fila existente : " + queueName);
        } catch (Exception e) {
            sqs.createQueue(CreateQueueRequest.builder().queueName(queueName).build());
            System.out.println("  [CRIADA] Fila criada    : " + queueName);
        }
        return queueUrl;
    }

    private static void sendMessage(SqsClient sqs, String queueUrl,
                                    MessageRequest payload, ObjectMapper mapper,
                                    String label) throws Exception {
        String body = mapper.writeValueAsString(payload);
        sqs.sendMessage(SendMessageRequest.builder()
                .queueUrl(queueUrl)
                .messageBody(body)
                .build());
        System.out.println("\n  [ENVIADO] " + label);
        System.out.println("  Fila  : " + queueUrl);
        System.out.println("  Body  : " + body.substring(0, Math.min(body.length(), 120)) + "...");
    }

    private static String buildQueueUrl(String queueName) {
        return LOCALSTACK_ENDPOINT + "/" + ACCOUNT_ID + "/" + queueName;
    }
}

File: test-scripts/send-test-messages.py
"""
Script de Teste - AsyncFlow
============================
Envia mensagens de teste para as filas SQS do LocalStack para testar:
  - validation-worker  (fila: validacao-vendas)
  - persistence-worker (fila: notas-processadas)

Pré-requisitos:
  pip install boto3
  Docker rodando com LocalStack (docker-compose up -d)

Uso:
  python send-test-messages.py
"""

import boto3
import json
import base64
from datetime import datetime

# ─────────────────────────────────────────────
# Configuração do LocalStack
# ─────────────────────────────────────────────
LOCALSTACK_ENDPOINT = "http://localhost:4566"
REGION = "us-east-1"
AWS_ACCESS_KEY = "test"
AWS_SECRET_KEY = "test"

# Nomes das filas
QUEUE_VALIDATION  = "validacao-vendas"    # validation-worker consome desta fila
QUEUE_PERSISTENCE = "notas-processadas"   # persistence-worker consome desta fila
QUEUE_MONITOR     = "fila-monitor"        # persistence-worker publica aqui

# ─────────────────────────────────────────────
# Cliente SQS
# ─────────────────────────────────────────────
sqs = boto3.client(
    "sqs",
    endpoint_url=LOCALSTACK_ENDPOINT,
    region_name=REGION,
    aws_access_key_id=AWS_ACCESS_KEY,
    aws_secret_access_key=AWS_SECRET_KEY,
)


def get_or_create_queue(queue_name: str) -> str:
    """Cria a fila se não existir e retorna a URL."""
    try:
        response = sqs.get_queue_url(QueueName=queue_name)
        url = response["QueueUrl"]
        print(f"[OK] Fila existente: {queue_name} -> {url}")
        return url
    except sqs.exceptions.QueueDoesNotExist:
        response = sqs.create_queue(QueueName=queue_name)
        url = response["QueueUrl"]
        print(f"[CRIADA] Fila criada: {queue_name} -> {url}")
        return url


def send_message(queue_url: str, payload: dict, label: str):
    """Envia uma mensagem JSON para a fila."""
    body = json.dumps(payload, ensure_ascii=False, default=str)
    sqs.send_message(QueueUrl=queue_url, MessageBody=body)
    print(f"\n[ENVIADO] {label}")
    print(f"  Fila : {queue_url}")
    print(f"  Body : {body[:200]}{'...' if len(body) > 200 else ''}")


# ─────────────────────────────────────────────
# Dados de Teste
# ─────────────────────────────────────────────

# Cliente de teste
CLIENTE_1 = {"cpf": "12345678901", "name": "João da Silva"}
CLIENTE_2 = {"cpf": "98765432100", "name": "Maria Oliveira"}

# Itens de teste
ITENS_1 = [
    {"serieNumber": "SN-001", "name": "Notebook Dell", "price": 3500.00},
    {"serieNumber": "SN-002", "name": "Mouse Logitech", "price": 150.00},
]
ITENS_2 = [
    {"serieNumber": "SN-003", "name": "Monitor LG 27\"", "price": 1200.00},
    {"serieNumber": "SN-004", "name": "Teclado Mecânico", "price": 450.00},
    {"serieNumber": "SN-005", "name": "Headset Sony",    "price": 320.00},
]

# Nota Fiscal 1 - para validation-worker (sem arquivo)
NOTA_1_VALIDATION = {
    "notaFiscal": {
        "numeroNota": "NF-2026-001",
        "client": CLIENTE_1,
        "totalValue": 3650.00,
        "itens": ITENS_1,
        "buyDate": datetime.now().isoformat(),
        "cnpj": "11222333000181",   # CNPJ válido para teste
        "status": "PROCESSANDO",
        "fileBase64": None,
        "fileName": None,
    }
}

# Nota Fiscal 2 - para validation-worker (com arquivo XML simulado)
XML_CONTENT = b"""<?xml version="1.0" encoding="UTF-8"?>
<NotaFiscal>
  <numero>NF-2026-002</numero>
  <emitente>
    <cnpj>11222333000181</cnpj>
    <nome>Empresa Teste LTDA</nome>
  </emitente>
  <destinatario>
    <cpf>98765432100</cpf>
    <nome>Maria Oliveira</nome>
  </destinatario>
  <itens>
    <item><serie>SN-003</serie><descricao>Monitor LG 27</descricao><valor>1200.00</valor></item>
    <item><serie>SN-004</serie><descricao>Teclado Mecanico</descricao><valor>450.00</valor></item>
    <item><serie>SN-005</serie><descricao>Headset Sony</descricao><valor>320.00</valor></item>
  </itens>
  <total>1970.00</total>
</NotaFiscal>"""

NOTA_2_VALIDATION = {
    "notaFiscal": {
        "numeroNota": "NF-2026-002",
        "client": CLIENTE_2,
        "totalValue": 1970.00,
        "itens": ITENS_2,
        "buyDate": datetime.now().isoformat(),
        "cnpj": "11222333000181",
        "status": "PROCESSANDO",
        "fileBase64": base64.b64encode(XML_CONTENT).decode("utf-8"),
        "fileName": "nf-2026-002.xml",
    }
}

# Nota Fiscal 1 - para persistence-worker (já validada)
NOTA_1_PERSISTENCE = {
    "notaFiscal": {
        "numeroNota": "NF-2026-001",
        "client": CLIENTE_1,
        "totalValue": 3650.00,
        "itens": ITENS_1,
        "buyDate": datetime.now().isoformat(),
        "cnpj": "11222333000181",
        "status": "VALIDADA",
        "fileBase64": None,
        "fileName": None,
    }
}

# Nota Fiscal 2 - para persistence-worker (já validada, com arquivo XML)
NOTA_2_PERSISTENCE = {
    "notaFiscal": {
        "numeroNota": "NF-2026-002",
        "client": CLIENTE_2,
        "totalValue": 1970.00,
        "itens": ITENS_2,
        "buyDate": datetime.now().isoformat(),
        "cnpj": "11222333000181",
        "status": "VALIDADA",
        "fileBase64": base64.b64encode(XML_CONTENT).decode("utf-8"),
        "fileName": "nf-2026-002.xml",
    }
}


# ─────────────────────────────────────────────
# Main
# ─────────────────────────────────────────────
def main():
    print("=" * 60)
    print("  AsyncFlow - Script de Teste de Mensagens SQS")
    print("=" * 60)

    # Garante que as filas existem
    print("\n[1] Verificando/criando filas no LocalStack...")
    url_validation  = get_or_create_queue(QUEUE_VALIDATION)
    url_persistence = get_or_create_queue(QUEUE_PERSISTENCE)
    url_monitor     = get_or_create_queue(QUEUE_MONITOR)

    # ── Mensagens para o VALIDATION-WORKER ──────────────────
    print("\n[2] Enviando mensagens para o VALIDATION-WORKER...")
    send_message(url_validation, NOTA_1_VALIDATION,
                 "NF-2026-001 (sem arquivo) -> validation-worker")
    send_message(url_validation, NOTA_2_VALIDATION,
                 "NF-2026-002 (com XML)     -> validation-worker")

    # ── Mensagens para o PERSISTENCE-WORKER ─────────────────
    print("\n[3] Enviando mensagens para o PERSISTENCE-WORKER...")
    send_message(url_persistence, NOTA_1_PERSISTENCE,
                 "NF-2026-001 (sem arquivo) -> persistence-worker")
    send_message(url_persistence, NOTA_2_PERSISTENCE,
                 "NF-2026-002 (com XML)     -> persistence-worker")

    print("\n" + "=" * 60)
    print("  Mensagens enviadas com sucesso!")
    print("  Verifique os logs dos workers para acompanhar o processamento.")
    print("=" * 60)


if __name__ == "__main__":
    main()
